src/prosemirror-view/viewdesc.ts

```diff
@@ -0,0 +1,1621 @@
+      domSel.collapse(anchorDOM.node, anchorDOM.offset)
+      try {
+        if (anchor != head)
+          domSel.extend(headDOM.node, headDOM.offset)
+        domSelExtended = true
+      } catch (_) {
+        // In some cases with Chrome the selection is empty after calling
+        // collapse, even when it should be valid. This appears to be a bug, but
+        // it is difficult to isolate. If this happens fallback to the old path
+        // without using extend.
+        // Similarly, this could crash on Safari if the editor is hidden, and
+        // there was no selection.
+      }
+    }
+    if (!domSelExtended) {
+      if (anchor > head) { let tmp = anchorDOM; anchorDOM = headDOM; headDOM = tmp }
+      let range = document.createRange()
+      range.setEnd(headDOM.node, headDOM.offset)
+      range.setStart(anchorDOM.node, anchorDOM.offset)
+      domSel.removeAllRanges()
+      domSel.addRange(range)
+    }
+  }
+
+  ignoreMutation(mutation: MutationRecord): boolean {
+    return !this.contentDOM && (mutation.type as any) != "selection"
+  }
+
+  get contentLost() {
+    return this.contentDOM && this.contentDOM != this.dom && !this.dom.contains(this.contentDOM)
+  }
+
+  // Remove a subtree of the element tree that has been touched
+  // by a DOM change, so that the next update will redraw it.
+  markDirty(from: number, to: number) {
+    for (let offset = 0, i = 0; i < this.children.length; i++) {
+      // @ts-expect-error
+      let child = this.children[i], end = offset + child.size
+      if (offset == end ? from <= end && to >= offset : from < end && to > offset) {
+        // @ts-expect-error
+        let startInside = offset + child.border, endInside = end - child.border
+        if (from >= startInside && to <= endInside) {
+          this.dirty = from == offset || to == end ? CONTENT_DIRTY : CHILD_DIRTY
+          if (from == startInside && to == endInside &&
+              // @ts-expect-error
+              (child.contentLost || child.dom.parentNode != this.contentDOM)) child.dirty = NODE_DIRTY
+          // @ts-expect-error
+          else child.markDirty(from - startInside, to - startInside)
+          return
+        } else {
+          // @ts-expect-error
+          child.dirty = child.dom == child.contentDOM && child.dom.parentNode == this.contentDOM && !child.children.length
+            ? CONTENT_DIRTY : NODE_DIRTY
+        }
+      }
+      offset = end
+    }
+    this.dirty = CONTENT_DIRTY
+  }
+
+  markParentsDirty() {
+    let level = 1
+    for (let node = this.parent; node; node = node.parent, level++) {
+      let dirty = level == 1 ? CONTENT_DIRTY : CHILD_DIRTY
+      if (node.dirty < dirty) node.dirty = dirty
+    }
+  }
+
+  get domAtom() { return false }
+
+  get ignoreForCoords() { return false }
+}
+
+// A widget desc represents a widget decoration, which is a DOM node
+// drawn between the document nodes.
+export class WidgetViewDesc extends ViewDesc {
```

Comment:
```suggestion
// $$FORK: WidgetViewDesc not currently exported from prosemirror-view
export class WidgetViewDesc extends ViewDesc {
```

Reply:
Good call. I actually need to take a minute at some point and confirm that either:

1) ProseMirror just works fine if you use custom ViewDesc classes that match the interface it cares about

OR

2) The existing ViewDesc classes will work for what we need.

I think that _both_ of these are likely to be true, so which one we go with will be more determined by how comfortable Marijn ends up feeling about exporting these classes. But again, I haven't verified yet.